### ISSUE TITLE
[JavaScript] typemap out is not used in constructors wrapper code

### ISSUE THREAD

**Issue by pizzafroide** [swig/swig#689]
```c++
%{
  #include <iostream>
  struct Foo {};
  Foo * bar() { return new Foo(); };
%}

%typemap(out) Foo * {
  std::cout << "typemap out" << std::endl;
  $typemap(out, SWIGTYPE *);
}

struct Foo
{
  %extend { Foo() { return new Foo(); }; }
};

%newobject bar;
Foo * bar();
```
When calling `bar()` from JavaScript, `"typemap out"` is printed (as expected). But when calling `Foo` constructor nothing is printed.

**Comment by wsfulton** [comment]
On 30 May 2016 at 10:44, pizzafroide <notifications@github.com> wrote:

> %{
>   #include <iostream>
>   struct Foo {};
>   Foo * bar() { return new Foo(); };
> %}
>
> %typemap(out) Foo * {
>   std::cout << "typemap out" << std::endl;
>   $typemap(out, SWIGTYPE *);
> }
> struct Foo
> {
>   %extend { Foo() { return new Foo(); }; }
> };
>
> %newobject bar;
> Foo * bar();
>
> When calling bar() from JavaScript, "typemap out" is printed (as
> expected). But when calling Foo constructor nothing is printed.
>
>
> That is rather surprising. The Javascript maintainers have gone awol at
the moment, but could you please raise this as a Github issue so that there
is a record of it as a bug.

Thanks
William

>

**Comment by wsfulton** [comment]
Agreed that is rather surprising. We really need an active Javascript maintainer to fix this and other Javascript bugs, but thanks for logging this for now.

**Comment by ojwb** [comment]
Still present in current git master.

**Comment by ojwb** [comment]
Still present in current git master.  Here's a simple script to aid testing:

```
#!/bin/sh
set -ex
${SWIG-../preinst-swig} -c++ ${TARGETLANG--javascript -v8} test.i
echo "This should find more than one use of the typemap"
grep 'typemap out' test_wrap.cxx
```

Good:

```
$ TARGETLANG=-php ./runme 
+ ../preinst-swig -c++ -php test.i
+ echo This should find more than one use of the typemap
This should find more than one use of the typemap
+ grep typemap out test_wrap.cxx
    std::cout << "typemap out" << std::endl;
    std::cout << "typemap out" << std::endl;
```

Bad:

```
$ ./runme 
+ ../preinst-swig -c++ -javascript -v8 test.i
+ echo This should find more than one use of the typemap
This should find more than one use of the typemap
+ grep typemap out test_wrap.cxx
    std::cout << "typemap out" << std::endl;
```

I tested all the other supported target languages, and this is also wrong for `-ruby`.

@mmomtchev Maybe you can take a look for Javascript?

**Comment by mmomtchev** [comment]
In fact, after looking in detail, this is a problem: is it expected that this typemap could return an object of a different type? This will break the language semantics, a constructor should always return an object of the constructed type.

In this particular example, the typemap will modify the constructor to return a generic `SWIGTYPE` pointer.

**Comment by mmomtchev** [comment]
The same goes for Python - `out` typemaps are not applied to constructors

**Comment by ojwb** [comment]
> In fact, after looking in detail, this is a problem: is it expected that this typemap could return an object of a different type? This will break the language semantics, a constructor should always return an object of the constructed type.

Good point, but I think that's just a bug in the provided example - AIUI the idea is that it wanting to add some code before expanding the existing typemap, but it seems we can't use `$typemap(out, Foo*)` here as trying to do so creates an expansion loop.  I think one needs to inline the code which `$typemap(out, Foo*)` would give, which is somewhat target-language-specific.  Here's a better testcase which looks suitable for at least Python and Javascript, and also tests and actual constructor as well as a constructor added via `%extend`:

```
%module test

%typemap(out) Foo * {
  std::cout << "typemap out" << std::endl;
  $result = SWIG_NewPointerObj(SWIG_as_voidptr($1), SWIGTYPE_p_Foo, SWIG_POINTER_NEW |  0 );
}

%extend Foo {
  Foo() { return new Foo(1); };
}

%newobject bar;

%inline %{

#include <iostream>
struct Foo {
  Foo(int) {}
};
Foo * bar() { return new Foo(1); };

%}
```

> The same goes for Python - `out` typemaps are not applied to constructors

They seem to be for me, both for the `%extend` constructor in the original testcase, and for an actual constructor - using the new testcase in `test.i`:

```
$ ../preinst-swig -c++ -python test.i
$ grep -C3 'typemap out' test_wrap.cxx 
  arg1 = static_cast< int >(val1);
  result = (Foo *)new Foo(arg1);
  {
    std::cout << "typemap out" << std::endl;
    resultobj = SWIG_NewPointerObj(SWIG_as_voidptr(result), SWIGTYPE_p_Foo, SWIG_POINTER_NEW |  0 );
  }
  return resultobj;
--
  if ((nobjs < 0) || (nobjs > 0)) SWIG_fail;
  result = (Foo *)new_Foo__SWIG_1();
  {
    std::cout << "typemap out" << std::endl;
    resultobj = SWIG_NewPointerObj(SWIG_as_voidptr(result), SWIGTYPE_p_Foo, SWIG_POINTER_NEW |  0 );
  }
  return resultobj;
--
  if (!SWIG_Python_UnpackTuple(args, "bar", 0, 0, 0)) SWIG_fail;
  result = (Foo *)bar();
  {
    std::cout << "typemap out" << std::endl;
    resultobj = SWIG_NewPointerObj(SWIG_as_voidptr(result), SWIGTYPE_p_Foo, SWIG_POINTER_NEW |  0 );
  }
  return resultobj;
```

**Comment by ojwb** [comment]
(PHP has problems with this situation too it seems since `$result` in the `out` typemap expands to `ZEND_THIS` so the typical `$result = ...` ends up assigning to `ZEND_THIS`, which doesn't compile.  I think `$result` needs to be a separate variable - will try to fix that.)

**Comment by mmomtchev** [comment]
In fact Python does include the typemap but still produces incorrect result. In Python, the class constructor is a global C method that is invoked by the Python wrapper where the real constructor is defined. So this test is impossible: https://github.com/swig/swig/compare/master...mmomtchev:swig:typemap_out_ctor

Maybe ideally, the out typemap of a constructor should simply ignore the returned value?

**Comment by mmomtchev** [comment]
Ok, after much tinkering, this is a huge problem for JavaScript:
In JavaScript, a constructor is invoked with `this` allocated and normally a constructor is not expected to return a value - it automatically returns `this` - just like C++. This means that it won't be possible to have typemaps that are compatible with both normal methods - where one has to create a new object in order to return it - and constructors - where `SWIG_NewPointerObj` is not to be called since the object constructed cannot be returned.

_However_,
in the ECMAScript 6 specification which defines JavaScript classes, it is possible to explicitly return any value from a constructor. This means that in theory it would be possible to return the value allocated by `SWIG_NewPointerObj` and to destroy the allocated `this` - this comes down to simply replacing the `this` pointer - but it breaks inheritance. A class that does this cannot be properly inherited as its constructor will be destructive. I don't think it is a good idea to design a feature around this which seems to be an ES6 specification oversight and not an intended JS feature.

The only solutions that I see to this problem are:
* To never share an `out` typemap between a constructor and a normal function - and to omit the `SWIG_NewPointerObj` from the constructor version
* To have a special JS constructor `out_ctor` typemap
* To automatically suppress the `SWIG_NewPointerObj` from the typemap

**Comment by ojwb** [comment]
I think we can reasonably treat an `out` typemap applied to a constructor which returns a different target language type to that which the constructor is for as not being a sensible case and so we shouldn't aim to support it.

It sounds like the PHP case is somewhat like the Javascript one - the `zval` for the PHP `this` is already allocated and the output of the typemap needs to be put into that rather than being the return value in PHP.  I haven't had a chance to poke at this yet, but I think for PHP at least `$result` can just be a temporary `zval` and then we transfer from that to `ZEND_THIS`.

So in the case where we're wrapping a constructor and there's an `out` typemap to apply, we need to generate some extra code to set up the temporary to point `$result` at, adjust what we replace `$result` with, and generate some extra code to copy the result into `ZEND_THIS`).
